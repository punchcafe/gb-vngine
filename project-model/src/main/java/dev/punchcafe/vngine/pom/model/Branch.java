package dev.punchcafe.vngine.pom.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Data;
import lombok.NoArgsConstructor;

@Data
@NoArgsConstructor
public class Branch {

    @JsonProperty("node-id")
    private String nodeId;
    // Use for either automatic or player based since it's just a yaml DTO
    @JsonProperty("predicate-expression")
    private String predicateExpression;
    private String prompt;
}
