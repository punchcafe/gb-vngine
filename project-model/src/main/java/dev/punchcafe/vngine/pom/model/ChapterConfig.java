package dev.punchcafe.vngine.pom.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Getter;

import java.util.List;
import java.util.Map;

@Getter
public class ChapterConfig {
    @JsonProperty("chapter-id")
    private String chapterId;
    private String firstNodeId;
    private List<Node> nodes;
    @JsonProperty("chapter-variables")
    private Map<String, VariableTypes> chapterVariables;
}
