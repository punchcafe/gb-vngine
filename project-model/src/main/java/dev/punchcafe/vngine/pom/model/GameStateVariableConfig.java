package dev.punchcafe.vngine.pom.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.Map;

@Getter
@Setter
@NoArgsConstructor
public class GameStateVariableConfig {
    @JsonProperty("game-state-variables")
    private Map<String, VariableTypes> gameStateVariables;
}
