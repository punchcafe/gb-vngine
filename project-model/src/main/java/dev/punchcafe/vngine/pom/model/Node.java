package dev.punchcafe.vngine.pom.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.util.List;

@NoArgsConstructor
@Setter
@Getter
public class Node {
    private String id;
    private NodeType type;
    @JsonProperty("narrative-id")
    private String narrativeId;
    @JsonProperty("game-state-modifiers")
    private List<String> gameStateModifiers;
    private List<Branch> branches;
}
