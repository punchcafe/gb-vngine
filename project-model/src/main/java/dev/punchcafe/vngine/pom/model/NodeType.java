package dev.punchcafe.vngine.pom.model;

public enum NodeType {
    AUTOMATIC, PLAYER
}
