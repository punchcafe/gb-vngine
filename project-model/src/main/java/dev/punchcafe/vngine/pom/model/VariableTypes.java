package dev.punchcafe.vngine.pom.model;

public enum VariableTypes {
    STR, BOOL, INT;
}
