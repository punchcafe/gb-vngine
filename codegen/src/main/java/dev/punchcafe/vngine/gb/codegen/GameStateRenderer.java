package dev.punchcafe.vngine.gb.codegen;

import dev.punchcafe.vngine.pom.model.GameStateVariableConfig;
import dev.punchcafe.vngine.pom.model.VariableTypes;
import lombok.Builder;
import lombok.Getter;

import java.util.Map;
import java.util.stream.Collectors;

@Builder
@Getter
public class GameStateRenderer {

    private static String DEFINITION_HEADERS = "" +
            "#ifndef GAMESTATE_TYPE_DEFINITION\n" +
            "#define GAMESTATE_TYPE_DEFINITION\n";
    private static String STRUCT_STATEMENT_OPENER = "struct GameState {\n";
    private static String STRUCT_STATEMENT_CLOSER = "};\n";
    private static String GLOBAL_GAMESTATE_TEMPLATE = "struct GameState %s;";

    private final String globalGameStateVariableName;
    private final int maxStringVariableLength;
    private final GameStateVariableConfig gameStateVariableConfig;

    public String render() {
        return new StringBuilder()
                .append(DEFINITION_HEADERS)
                .append(STRUCT_STATEMENT_OPENER)
                .append(this.renderGameStateContents())
                .append(STRUCT_STATEMENT_CLOSER)
                .append(this.renderGameStateSingleton())
                .toString();
    }

    private String renderGameStateContents() {
        return this.gameStateVariableConfig
                .getGameStateVariables()
                .entrySet()
                .stream()
                .map(this::renderVariableOfType)
                .map(variable -> RenderUtils.addIndentationOfLevel(variable, 1))
                .sorted()
                .collect(Collectors.joining("\n"));
    }

    private String renderVariableOfType(final Map.Entry<String, VariableTypes> variableEntry) {
        switch (variableEntry.getValue()) {
            case BOOL:
                return String.format("short %s;", variableEntry.getKey());
            case STR:
                return String.format("char %s [%d];", variableEntry.getKey(), this.maxStringVariableLength);
            case INT:
                return String.format("int %s;", variableEntry.getKey());
            default:
                throw new UnsupportedOperationException();
        }
    }

    private String renderGameStateSingleton() {
        return String.format(GLOBAL_GAMESTATE_TEMPLATE, this.globalGameStateVariableName);
    }
    /*
    TODO:
    consider a List<Class<? extends Renderer>> getDependencies() for rendering everything into one big c file in the right order?
     */
}
