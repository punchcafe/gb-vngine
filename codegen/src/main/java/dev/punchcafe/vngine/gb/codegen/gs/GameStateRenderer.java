package dev.punchcafe.vngine.gb.codegen.gs;

import dev.punchcafe.vngine.gb.codegen.ComponentRenderer;
import dev.punchcafe.vngine.gb.codegen.RenderUtils;
import dev.punchcafe.vngine.gb.codegen.csan.CVariableNameSanitiser;
import dev.punchcafe.vngine.pom.model.GameStateVariableConfig;
import dev.punchcafe.vngine.pom.model.VariableTypes;
import dev.punchcafe.vngine.pom.model.vngpl.variable.GameVariableLevel;
import lombok.Builder;
import lombok.Getter;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@Builder
@Getter
public class GameStateRenderer implements ComponentRenderer {

    public static String GAME_STATE_RENDERER_NAME = "GAME_STATE_RENDERER_NAME";

    private static String DEFINITION_HEADERS = "" +
            "#ifndef GAMESTATE_TYPE_DEFINITION\n" +
            "#define GAMESTATE_TYPE_DEFINITION\n";

    private static String DEFINITION_FOOTER = "\n#endif";
    private static String STRUCT_STATEMENT_OPENER = "struct GameState {\n";
    private static String STRUCT_STATEMENT_CLOSER = "};\n";
    private static String GLOBAL_GAMESTATE_TEMPLATE = "struct GameState %s;";

    private final String globalGameStateVariableName;
    private final int maxStringVariableLength;
    private final GameStateVariableConfig gameStateVariableConfig;

    public String render() {
        return new StringBuilder()
                .append(DEFINITION_HEADERS)
                .append(STRUCT_STATEMENT_OPENER)
                .append(this.renderGameStateContents())
                .append(STRUCT_STATEMENT_CLOSER)
                .append(this.renderGameStateSingleton())
                .append(DEFINITION_FOOTER)
                .toString();
    }

    @Override
    public List<String> dependencies() {
        return List.of();
    }

    @Override
    public String componentName() {
        return GAME_STATE_RENDERER_NAME;
    }

    private String renderGameStateContents() {
        return this.gameStateVariableConfig
                .getGameStateVariables()
                .entrySet()
                .stream()
                .map(this::renderVariableOfType)
                .map(variable -> RenderUtils.addIndentationOfLevel(variable, 1))
                .sorted()
                .collect(Collectors.joining("\n"));
    }

    private String renderVariableOfType(final Map.Entry<String, VariableTypes> variableEntry) {
        final var variableName = CVariableNameSanitiser.sanitiseVariableName(variableEntry.getKey(),
                variableEntry.getValue(),
                GameVariableLevel.GAME);
        switch (variableEntry.getValue()) {
            case BOOL:
                return String.format("short %s;", variableName);
            case STR:
                return String.format("char %s [%d];", variableName, this.maxStringVariableLength);
            case INT:
                return String.format("int %s;", variableName);
            default:
                throw new UnsupportedOperationException();
        }
    }

    private String renderGameStateSingleton() {
        return String.format(GLOBAL_GAMESTATE_TEMPLATE, this.globalGameStateVariableName);
    }
}
