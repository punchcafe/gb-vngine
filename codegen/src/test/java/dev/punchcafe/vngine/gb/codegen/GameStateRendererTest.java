package dev.punchcafe.vngine.gb.codegen;

import dev.punchcafe.vngine.gb.codegen.render.gs.GameStateRenderer;
import dev.punchcafe.vngine.pom.model.GameStateVariableConfig;
import dev.punchcafe.vngine.pom.model.VariableTypes;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import java.util.Map;

import static io.github.jsonSnapshot.SnapshotMatcher.expect;
import static io.github.jsonSnapshot.SnapshotMatcher.start;

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
public class GameStateRendererTest {

    @BeforeAll
    void beforeAll() {
        start(Snapshot::asJsonString);
    }

    @Test
    void shouldRenderSimpleGameState() {
        final var config = new GameStateVariableConfig();
        final var params = Map.of("int-property-1", VariableTypes.INT,
                "int-property-2", VariableTypes.INT,
                "str-property-1", VariableTypes.STR,
                "str-property-2", VariableTypes.STR,
                "bool-property-1", VariableTypes.BOOL,
                "bool-property-2", VariableTypes.BOOL);
        config.setGameStateVariables(params);
        final var renderer = GameStateRenderer.builder()
                .gameStateVariableConfig(config)
                .globalGameStateVariableName("GAME_STATE_VARIABLE")
                .maxStringVariableLength(50)
                .build();
        expect(renderer.render()).toMatchSnapshot();
    }
}
