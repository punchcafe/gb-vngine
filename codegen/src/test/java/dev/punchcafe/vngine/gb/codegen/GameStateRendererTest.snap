dev.punchcafe.vngine.gb.codegen.GameStateRendererTest.shouldRenderSimpleGameState=[
  "#ifndef GAMESTATE_TYPE_DEFINITION\n#define GAMESTATE_TYPE_DEFINITION\nstruct GameState {\n    char str-property-1 [50];\n    char str-property-2 [50];\n    int int-property-1;\n    int int-property-2;\n    short bool-property-1;\n    short bool-property-2;};\nstruct GameState GAME_STATE_VARIABLE;"
]